tests/integration/porter/control/test_porter_web_control_federated.py

```diff
@@ -80,52 +84,96 @@ def test_get_ursulas(federated_porter_web_controller, federated_ursulas):
+                       b'he is (they are) being educated.'  # - James Baldwin
+    retrieve_cfrags_params, message_kit = retrieval_request_setup(enacted_federated_policy,
+                                                                  federated_bob,
+                                                                  federated_alice,
+                                                                  original_message=original_message,
+                                                                  encode_for_rest=True)
 
+    #
     # Success
-    exec_work_order_params = {
-        'ursula': ursula_address,
-        'work_order_payload': work_order_payload_b64
-    }
-    response = federated_porter_web_controller.post('/exec_work_order', data=json.dumps(exec_work_order_params))
+    #
+    response = federated_porter_web_controller.post('/retrieve_cfrags', data=json.dumps(retrieve_cfrags_params))
     assert response.status_code == 200
 
     response_data = json.loads(response.data)
-    work_order_result = response_data['result']['work_order_result']
-    assert work_order_result
+    retrieval_results = response_data['result']['retrieval_results']
+    assert retrieval_results
+
+    # expected results - can only compare length of results, ursulas are randomized to obtain cfrags
+    retrieve_args = retrieval_params_decode_from_rest(retrieve_cfrags_params)
+    expected_results = federated_porter.retrieve_cfrags(**retrieve_args)
+    assert len(retrieval_results) == len(expected_results)
+
+    # check that the re-encryption performed was valid
+    treasure_map = retrieve_args['treasure_map']
+    policy_message_kit = message_kit.as_policy_kit(policy_key=enacted_federated_policy.public_key,
+                                                   threshold=treasure_map.threshold)
+    assert len(retrieval_results) == 1
+    field = RetrievalResultSchema()
+    cfrags = field.load(retrieval_results[0])['cfrags']
+    verified_cfrags = {}
+    for ursula, cfrag in cfrags.items():
+        # need to obtain verified cfrags (verified cfrags are not deserializable, only non-verified cfrags)
+        verified_cfrag = cfrag.verify(capsule=policy_message_kit.message_kit.capsule,
+                                      verifying_pk=federated_alice.stamp.as_umbral_pubkey(),
+                                      delegating_pk=enacted_federated_policy.public_key,
+                                      receiving_pk=federated_bob.public_keys(DecryptingPower))
+        verified_cfrags[ursula] = verified_cfrag
+    retrieval_result_object = RetrievalResult(cfrags=verified_cfrags)
+    policy_message_kit = policy_message_kit.with_result(retrieval_result_object)
+
+    assert policy_message_kit.is_decryptable_by_receiver()
+
+    cleartext_with_sig_header = federated_bob._crypto_power.power_ups(DecryptingPower).keypair.decrypt(policy_message_kit)
+    sig_header, remainder = default_constant_splitter(cleartext_with_sig_header, return_remainder=True)
```

Comment:
This should really be fixed as a part of #2743. The signature header (if any) should be separated on decryption.

Reply:
👍, I was basically mimicking the decryption code from here - https://github.com/nucypher/nucypher/blob/main/nucypher/characters/base.py#L426. Didn't want to use Bob's Python API to decrypt the result, to better mimic how Porter would actually be used, and ensure that it worked - https://github.com/nucypher/nucypher/pull/2768/commits/7f20fbdc7bc7a9a0aff6c3a2714dacc672bd3008.